scipy/sparse/_sputils.py

```diff
@@ -297,7 +301,8 @@ def validateaxis(axis) -> None:
         raise ValueError("axis out of range")
 
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
+                allow_nd=False)
```

Comment:
Can we combine the parameters `allow_1d` and `allow_nd`? I believe they always take on the value `is_array` which just checks for sparray (and thus not spmatrix). As far as shapes are concerned, sparrays should allow nD. Specific formats may not support nD, but that is checked elsewhere -- no check of the shape should depend on the format. 

Some background: When we added support for 1D arrays we first wrote it using the parameter name `allow_nd`. We changed it to `allow_1d` when we delayed support for nd.

There doesn't seem to be any need for two parameters in `isshape` and in `check_shape`. The long term solution is to have `allow_nd` carry this information.  I would be fine with making the switch. But if people want a deprecation cycle, that should work too. Keep the old `allow_1d`, add the `allow_nd` and at the beginning of the functions check whether `allow_1d` is not its default, in which case raise a DeprecationWarning and set  `allow_nd` to the value given. Otherwise just use `allow_nd`. After two releases the deprecation can be backed out.

Do you have thoughts about using both parameters over just using `allow_nd`?

Reply:
The primary concern for having both separately is because other formats would allow 1d but now allow nd, whereas for COO we would allow both. This would not break any existing 1D stuff for other formats, and if someone in the future wants to extend other formats such as CSR to n-D, that can be done easily by setting `allow_nd` to `is_array` instead of letting it default to `False`.
If we were to combine `allow_1d` and `allow_nd`, I'm a bit concerned about the handling of, example, the following code from [_sputils.py](https://github.com/anushkasuyal/scipy/blob/f36e52d28582007619d9717cd9f8924694447118/scipy/sparse/_sputils.py#L343-L351) -
```py
    if current_shape is None:
        if not allow_1d and len(new_shape) != 2:
            raise ValueError('shape must be a 2-tuple of positive integers')

        if not allow_nd and len(new_shape) > 2:
            if allow_1d:
                raise ValueError('shape must be a 1- or 2-tuple of positive integers')
            else:
                raise ValueError('shape must be a 2-tuple of positive integers')
```
We do want to raise the "shape must be a 1- or 2-tuple of positive integers" error for other formats, (which allow 1d but not nD), but for COO there is no such case.

Reply:
`allow_1d` and `allow_nd` have now been combined into a tuple-valued parameter, `allowed_ndims`, at https://github.com/scipy/scipy/commit/0ea748b03fb949919287bc9f57ba0e0df0fdad4a.